File: Sources/ClaudeNotch/Currency.swift
import Foundation
import Combine

/// 金额显示：内部一律存**美元**（LiteLLM / 覆盖价都是 USD），仅在显示时按当前语言折算。
/// 英语 → `$1.23`；中文 → `¥8.79`（= USD × 实时汇率）。
func money(_ usd: Double, decimals: Int = 2) -> String {
    if LocalizationState.currentLanguage == .zh {
        return "¥" + String(format: "%.\(decimals)f", usd * LocalizationState.usdToCny)
    }
    return "$" + String(format: "%.\(decimals)f", usd)
}

/// 带「≈」前缀的金额（订阅用户为等价折算，非真实扣费，中英文都标）。
func approxMoney(_ usd: Double, decimals: Int = 2) -> String { "≈" + money(usd, decimals: decimals) }

// MARK: - 汇率（USD→CNY）：内置默认 + 联网每周刷新 + 手动刷新

@MainActor
final class ExchangeRateStore: ObservableObject {
    /// 免费公开汇率 API（无需认证）。
    static let remoteURL = URL(string: "https://open.er-api.com/v6/latest/USD")!
    private static let refreshInterval: TimeInterval = 7 * 24 * 3600
    private static let defaultRate = 7.15
    private static let rateKey = "usdToCnyRate"
    private static let fetchedAtKey = "usdToCnyFetchedAt"

    @Published private(set) var rate: Double = ExchangeRateStore.defaultRate
    @Published private(set) var lastUpdated: Date?     // 联网成功时刻；nil = 仅默认值
    @Published private(set) var isRefreshing = false
    @Published private(set) var lastError: String?

    /// 启动：先用缓存（或默认）即时生效，再按需后台刷新（缺缓存或超 7 天）。
    func bootstrap() {
        let d = UserDefaults.standard
        let cached = d.object(forKey: Self.rateKey) as? Double
        let fetchedAt = d.object(forKey: Self.fetchedAtKey) as? Date
        let r = cached ?? Self.defaultRate
        rate = r
        lastUpdated = fetchedAt
        LocalizationState.usdToCny = r
        let stale = fetchedAt.map { Date().timeIntervalSince($0) > Self.refreshInterval } ?? true
        if stale { refresh() }
    }

    /// 手动 / 自动刷新：联网取 USD→CNY → 落盘 → 即时生效。
    func refresh() {
        if isRefreshing { return }
        isRefreshing = true
        lastError = nil
        var req = URLRequest(url: Self.remoteURL)
        req.timeoutInterval = 15
        URLSession.shared.dataTask(with: req) { [weak self] data, _, err in
            let cny: Double? = {
                guard let data,
                      let root = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any],
                      (root["result"] as? String) == "success",
                      let rates = root["rates"] as? [String: Any],
                      let v = (rates["CNY"] as? NSNumber)?.doubleValue, v > 0 else { return nil }
                return v
            }()
            DispatchQueue.main.async {
                guard let self else { return }
                self.isRefreshing = false
                if let cny {
                    let now = Date()
                    self.rate = cny
                    self.lastUpdated = now
                    LocalizationState.usdToCny = cny
                    UserDefaults.standard.set(cny, forKey: Self.rateKey)
                    UserDefaults.standard.set(now, forKey: Self.fetchedAtKey)
                } else {
                    self.lastError = err?.localizedDescription ?? tr("汇率解析失败", "Failed to parse exchange rate")
                }
            }
        }.resume()
    }
}

File: windows/ClaudeNotch/Core/Currency.cs
using System.Globalization;
using System.Net.Http;
using System.Text.Json;

namespace ClaudeNotch.Core;

/// <summary>金额显示：内部一律存美元，显示时英文 $ / 中文 ¥(×实时汇率)。</summary>
public static class Money
{
    public static double UsdToCny = 7.15;   // 离线默认，联网成功后覆盖

    public static string Format(double usd, int decimals = 2)
    {
        var ci = CultureInfo.InvariantCulture;
        return L.Current == AppLang.Zh
            ? "¥" + (usd * UsdToCny).ToString("F" + decimals, ci)
            : "$" + usd.ToString("F" + decimals, ci);
    }

    public static string Approx(double usd, int decimals = 2) => "≈" + Format(usd, decimals);
}

/// <summary>USD→CNY 汇率：内置默认 + 联网每周刷新（open.er-api.com 公开数据，无需认证）+ 手动刷新。</summary>
public sealed class ExchangeRateStore
{
    const string RemoteUrl = "https://open.er-api.com/v6/latest/USD";
    static readonly TimeSpan RefreshInterval = TimeSpan.FromDays(7);
    const double DefaultRate = 7.15;

    public double Rate { get; private set; } = DefaultRate;
    public DateTime? LastUpdated { get; private set; }
    public bool IsRefreshing { get; private set; }
    public string? LastError { get; private set; }
    public event Action? Changed;

    static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(15) };

    string CacheFile => Path.Combine(Paths.SupportDir, "exchange-rate.json");

    record Cache(double rate, long fetchedAtUnix);

    public void Bootstrap()
    {
        double r = DefaultRate;
        DateTime? fetchedAt = null;
        try
        {
            if (File.Exists(CacheFile))
            {
                var c = JsonSerializer.Deserialize<Cache>(File.ReadAllText(CacheFile));
                if (c is not null && c.rate > 0)
                {
                    r = c.rate;
                    fetchedAt = DateTimeOffset.FromUnixTimeSeconds(c.fetchedAtUnix).LocalDateTime;
                }
            }
        }
        catch { /* 用默认 */ }

        Rate = r;
        Money.UsdToCny = r;
        LastUpdated = fetchedAt;
        Changed?.Invoke();

        bool stale = fetchedAt is null || DateTime.Now - fetchedAt.Value > RefreshInterval;
        if (stale) _ = RefreshAsync();
    }

    public async Task RefreshAsync()
    {
        if (IsRefreshing) return;
        IsRefreshing = true; LastError = null; Changed?.Invoke();
        try
        {
            var json = await Http.GetStringAsync(RemoteUrl).ConfigureAwait(false);
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;
            if (root.TryGetProperty("result", out var res) && res.GetString() == "success"
                && root.TryGetProperty("rates", out var rates)
                && rates.TryGetProperty("CNY", out var cny) && cny.TryGetDouble(out var v) && v > 0)
            {
                Rate = v;
                Money.UsdToCny = v;
                LastUpdated = DateTime.Now;
                File.WriteAllText(CacheFile, JsonSerializer.Serialize(
                    new Cache(v, DateTimeOffset.UtcNow.ToUnixTimeSeconds())));
            }
            else LastError = L.Tr("汇率解析失败", "Failed to parse exchange rate");
        }
        catch (Exception e) { LastError = e.Message; }
        finally { IsRefreshing = false; Changed?.Invoke(); }
    }
}
